File: Planetary-Formation-main/Storage/STL/StlFile.java
package Storage.STL;

import java.io.FileWriter;
import java.io.IOException;

public class StlFile {
    private FileWriter file;

    private int triCount;

    public int getTriCount() {
        return triCount;
    }

    // Header is 80B, 10B reserved for unit declaration, 70B remain
    // Reserve 4B at the start of the file for triangle count
    // Creates a file interface with the filename, the unit that the file will be stored in, and a short informative header.
    StlFile(String filename, String unit, String header) throws IOException {
        file = new FileWriter(filename);
        triCount = 0;

        if (header.length() != 72)
            throw new IllegalArgumentException("Header must be 70 bytes");
        if (unit.length() != 2)
            throw new IllegalArgumentException("Unit must be 2 bytes (mm,cm, m,ft,in,ly)");

        file.write(header + "\nUNITS=" + unit + "\n    ");
    }

    // Writes a TriangleFace to the file
    public void writeTriangle(TriangleFace triangle) throws IOException {
        this.triCount ++;

        byte[] code = triangle.exportSTLCode();

        for (byte codeByte : code)
            this.file.append((char) codeByte);
    }
}

File: Planetary-Formation-main/storage/STL/StlFile.kt
package storage.STL

import java.io.FileWriter
import java.io.IOException

class StlFile constructor(filename: String, unit: String, header: String) {
    private val file: FileWriter
    private var triCount: Int
    private val buffer: CharArray
    private var bufferUsage: Int

    fun getTriCount(): Int {
        return triCount
    }

    // Header is 80B, 10B reserved for unit declaration, 70B remain
    // Reserve 4B at the start of the file for triangle count
    init {
        file = FileWriter(filename)
        triCount = 0
        buffer = CharArray(65536)
        bufferUsage = 0
        if (header.length > 70) throw IllegalArgumentException("Header must be 70 bytes or less")
        if (unit.length != 2) throw IllegalArgumentException("Unit must be 2 bytes (mm,cm, m,ft,in,ly)")
        if (header.length < 70) {
            file.write("$header\nUNITS=$unit" +
                    String(CharArray(70-header.length) { return@CharArray ' ' }) +
                    "\n    ")
        } else {
            file.write("$header\nUNITS=$unit\n    ")//TODO Write to triangle count
        }
    }

    @Throws(IOException::class)
    fun writeTriangle(triangle: TriangleFace) {
        triCount++
        val code: ByteArray = triangle.exportCode()
        for (codeByte: Byte in code) buffer[bufferUsage] = Char(codeByte.toUShort())
        bufferUsage++
        if (bufferUsage == 65536) {
            file.append(buffer.toString())
            bufferUsage = 0
        }
    }

    @Throws(IOException::class)
    fun writeTriangles(triangles: Array<TriangleFace>) {
        for (triangle: TriangleFace in triangles) writeTriangle(triangle)
    }

    fun close() {
        val remainingBuffer: CharArray = CharArray(bufferUsage) { i: Int -> return@CharArray buffer[i] }
        file.append(remainingBuffer.toString())
        file.close()
    }
}